### PR TITLE
Adds variable Cd and Manning's n values 

### PR DESCRIPTION
**Pull request by pwolfram**
Adds variable bottom drag capability as well as Manning's n values
for variable drag in coastal areas.

Note, equation for Manning's n conversion to Cd is `Cd=g*n^2*h^(-1/3)`, where `n` is Manning's coefficient, `g` is gravity, and `h` is the total water depth.

Testing is against the drying slope case.

### CODE REVIEW

**Comment by pwolfram** [comment]
Note, need to remove RK4 fixes by Michael and need to squash commits.  Can do this after folks have a chance to review the code.

**Comment by pwolfram** [comment]
# Testing

## variable Cd: `ocean/drying_slope/zstar_variableCd/1km`
dryingslopecomparison

## variable Manning's n: `ocean/drying_slope/zstar_variableManningn/1km`
dryingslopecomparisonManningn

**Comment by pwolfram** [comment]
Could potentially use a function pointer to make things simpler @mark-petersen but kept standard approach we've been using and duplicated function but with need high-level key terms to keep functionality with switch outside any main loops for performance reasons.

**Comment by pwolfram** [comment]
@mark-petersen, can you please review?  The "Don't Merge" label is just to point out that I have Michael's RK4 fix (needs removed since that will be merged into dev) and need to squash commits prior to the merge.

File: src/core_ocean/shared/mpas_ocn_vmix.F [hunk 1]
```diff
@@ -481,6 +482,322 @@ subroutine ocn_vel_vmix_tend_implicit(meshPool, dt, kineticEnergyCell, vertViscT
...
+         ! so recompute layerThicknessEdge here.
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+         do k = 1, N
+            layerThicknessEdge(k,iEdge) = 0.5_RKIND * (layerThickness(k,cell1) + layerThickness(k,cell2))
+         end do
+
+         ! average cell-based implicit bottom drag to edges and convert Mannings n to Cd
+         implicitCd = gravity*(0.5_RKIND*(bottomDrag(cell1) + bottomDrag(cell2)))**2.0 * &
+           0.5_RKIND * (ssh(cell1) + ssh(cell2) + bottomDepth(cell1) + bottomDepth(cell2))**(-1.0_RKIND/3.0_RKIND)
```
Comment (mark-petersen) [inline src/core_ocean/shared/mpas_ocn_vmix.F]: It does seem unfortunate to add two new subroutines here to alter two lines, the calculation of implicitCd and for B(n) below.  This goes back to design choices of efficiency versus versatility. Choices are:
1. Extra subroutines with redundant code (your choice here).
2. If statements around smaller loops.
3. Include extra calculations with 0/1 coefficients for flag choices.
4. C pre-processor branching

None are perfect, but I agree with your choice here.

**Comment by mark-petersen** [comment]
@pwolfram Please run your standard cases to exercise this code with gnu and intel compilers, using DEBUG=true.

File: src/core_ocean/shared/mpas_ocn_vmix.F (hunk 1 above)
Comment (pwolfram) [reply src/core_ocean/shared/mpas_ocn_vmix.F]: Thanks Mark!

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F [hunk 2]
```diff
@@ -244,6 +252,15 @@ subroutine ocn_init_setup_tidal_boundary(domain, iErr)!{{{
           (config_tidal_boundary_right_bottom_depth - config_tidal_boundary_left_bottom_depth)
       end do
 
+      if (config_use_variable_drag) then
```
Comment (pwolfram) [inline src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F]: @caozd999, here is the selection of different types of bottom drag you can use as an example to build your own transect and variable drag.

Comment (pwolfram) [reply src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F]: https://github.com/MPAS-Dev/MPAS-Model/pull/389/files#diff-6270014bc9ae9b8554f8af50127caf5fL242 above shows bottom depth that needs modified.

I recommend building your own new `src/core_ocean/mode_init/mpas_ocn_init_tidal_transect.F` case.

**Comment by pwolfram** [comment]
@mark-petersen, here are the tests you've requested.

# Testing

## intel-debug
dryingslopecomparisonManningn

## intel-release
dryingslopecomparisonManningn

Gnu to follow later...

**Comment by pwolfram** [comment]
# Testing

## gnu-release
dryingslopecomparisonManningn

## gnu-debug
dryingslopecomparisonManningn

**Comment by pwolfram** [comment]
@mark-petersen and @sbrus89, I'm planning to merge because everything checks out.  If you both could approve before I do this I would appreciate it.

**Comment by sbrus89** [comment]
@pwolfram, the only comment I have would be that the existing constant bottom drag could be fit into this framework by just setting the `bottomDrag` variable to that constant. It adds an additional array for runs that don't require the spatially variable functionality, but would keep things simpler because there would be one fewer `ocn_vel_vmix_tend_implicit` routines. 

Just a thought and definitely not an issue that should be addressed by this PR.

**Comment by pwolfram** [comment]
Thanks @sbrus89!  I'll let @mark-petersen comment here, but believe we should skip that because it would modify the E3SM runtime and not necessarily be BFB.  Minimally, it would change performance for large-scale runs with constant Cd.  But, I agree, we are adding complexity via leaving that additional option in, which your approach would remove.

**Comment by pwolfram** [comment]
Thanks @mark-petersen, double-checked it worked on the merge commit I just pushed.  Thanks also @sbrus89

**Comment by pwolfram** [comment]
@caozd999, PR is now merged.

**Comment by caozd999** [comment]
@pwolfram great! Thank you.

File: src/core_ocean/shared/mpas_ocn_vmix.F [hunk 3]
```diff
@@ -481,6 +482,322 @@ subroutine ocn_vel_vmix_tend_implicit(meshPool, dt, kineticEnergyCell, vertViscT
...
+         ! layerThicknessEdge is computed in compute_solve_diag, and is not available yet,
+         ! so recompute layerThicknessEdge here.
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+         do k = 1, N
+            layerThicknessEdge(k,iEdge) = 0.5_RKIND * (layerThickness(k,cell1) + layerThickness(k,cell2))
+         end do
+
+         ! average cell-based implicit bottom drag to edges and convert Mannings n to Cd
+         implicitCd = gravity*(0.5_RKIND*(bottomDrag(cell1) + bottomDrag(cell2)))**2.0 * &
```
Comment (pwolfram) [inline src/core_ocean/shared/mpas_ocn_vmix.F]: @caozd999, here is the example of how I converted Manning's n into Cd for computation in the momentum equations.

**Pull request merged**